### DISCUSSION TITLE
Expose i18n data in DocumentContext

### DISCUSSION THREAD

**Discussion by santialbo** [vercel/next.js#18329] [Ideas]
# Feature request
Right now (trying out official i18n support) I'm using a custom `_document` file to load the translations files I need and making some small changes into the generated document so this data can be rehydrated on the client.

I was expecting the selected locale to be part of the `DocumentContext` passed to this custom document's `getInitialProps` but it's not there.

## Describe the solution you'd like

Pass the selected locale (and possibly any i18n config data) to the custom document's `getInitialProps`.

**Comment by timneutkens** [comment]
Can you provide more details about the "why" you want this change. What's the thing you're trying to achieve?

**Reply by santialbo** [reply]
I hope I'm able to explain what we are doing. I got very excited when you announced the native i18n support because it's very similar to what we have been doing so far.

What I'm trying to achieve is the same we have running right now but using the native i18n support. For that to happen we need to access the selected locale in `Document.getInitialProps`.

Let me give you some context of what we are doing right now on (v.9.5.x).

We are using a path based strategy for i18n (we have 2 languages, english `en` and spanish `es`) where our file structure looks like this
```
- pages
  - index.js // redirects based on `navigator.languages` on the client or the `Accept-Language` header on the server
  - [locale]
    - index.js
    - xxxxx.js
```
In our custom `Document.getInitialProps` (code snippet below) we get the locale from the query and load the i18n data.
We pass that data to our custom `_app` that (on the server) will use it to pass it to the `I18nProvider`.

We also include the following line ```<script src={`/static/lang/${this.props.locale}.js`} />``` which (on the client) will run a code that looks like `window.__LOCALE_DATA__ = {....}` and which will allow our custom `_app` to correctly rehydrate the client by passing it to the `I18nProvider`.

We chose to follow this approach over using `App.getInitialProps` to avoid bloating the generated html with all the locale data in them and have, instead, a single cacheable file that will be used on all pages.

```ts
// simplified _document.ts
class MyDocument extends Document<MyDocumentProps> {
  static async getInitialProps(ctx: DocumentContext) {
    const locale = ctx.query.locale as string; // this is the missing part on v10. there's no "locale" query parameter
    const messages = await loadMessages(locale);
    const { renderPage } = ctx;
    ctx.renderPage = () => renderPage({
      enhanceApp: (App) => (props) => (<App {...props} {...{ locale, messages }} />),
    });
    const initialProps = await Document.getInitialProps(ctx);
    return { ...initialProps, locale };
  }

  render() {
    return (
      <Html lang={this.props.locale}>
        <Head>
          {...}
        </Head>
        <body>
          <Main />
          <script src={`/static/lang/${this.props.locale}.js`} />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```

```ts
// simplified _app.ts
function MyApp({ Component, pageProps, router, ...props }: MyAppProps) {
  const locale = router.query.locale;
  const messages = process.browser
    ? window.__LOCALE_DATA__ // runs on the client
    : props.messages; // runs on the browser, passed by the custom _document
  return (
    <I18nProvider locale={locale} messages={messages}>
      <Component {...pageProps} />
    </I18nProvider>
  );
}
```

I hope all of this makes a bit of sense. If something is not clear enough please let me know and I'll try to explain it a bit more in detail.

Thanks!

**Reply by maktouch** [reply]
Is there another way to get the locale in _document, so that we can pass it to the Html tag?

**Reply by IskanderMustafin** [reply]
> Can you provide more details about the "why" you want this change. What's the thing you're trying to achieve?

In my case I need to set `dir` attribute in `<Html/>` component (`i18n` feature take care only about `lang` attribute). So I need access to current locale to determine if it is `rtl` or `ltr`.

Also, as shown in next.js with material-ui [example](https://github.com/mui-org/material-ui/blob/next/examples/nextjs/pages/_document.js), we need to create mui theme inside  Document. But again, in my case `mui-theme` is created dynamically according to current `direction`: 

```
const langDir: 'rtl' | 'ltr' = getLangDirByLocale(locale); // I need current locale here (inside _document.tsx)
const muiTheme = createMuiTheme(langDir);
```

**Reply by santialbo** [reply]
Hi @timneutkens, what are your thoughts on these reasons?

**Reply by hedwiggggg** [reply]
Posted a workaround: https://github.com/vercel/next.js/discussions/18396#discussioncomment-387969
Maybe this helps somebody :)

**Comment by santialbo** [comment]
I hate to be the person doing this but, can we get an answer from the next team? @timneutkens @Timer
Is this something you would accept PRs for?
If so would it be OK just including `locale`, `defaultLocale` and `locales` on the `DocumentContext`?

**Reply by timneutkens** [reply]
We're planning to add translation loading which would cover this much better than exposing these values in getInitialProps of _document.

**Reply by maktouch** [reply]
That will kinda solve the problem, but not fully. By not making it available in _document, we won't have any transition path..

**Reply by santialbo** [reply]
That sounds very interesting @timneutkens. Is there an RFC for this upcoming feature?
